specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+# Custom Gas Token
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+- [Constants](#constants)
+- [Properties of a Gas Paying Token](#properties-of-a-gas-paying-token)
+- [Configuring the Gas Paying Token](#configuring-the-gas-paying-token)
+- [Contract Modifications](#contract-modifications)
+  - [OptimismPortal](#optimismportal)
+    - [`depositERC20Transaction`](#depositerc20transaction)
+    - [`depositTransaction`](#deposittransaction)
+  - [StandardBridge](#standardbridge)
+  - [CrossDomainMessenger](#crossdomainmessenger)
+  - [L1Block](#l1block)
+- [User Flow](#user-flow)
+- [Upgrade](#upgrade)
+  - [L1Block Deployment](#l1block-deployment)
+  - [L2CrossDomainMessenger Deployment](#l2crossdomainmessenger-deployment)
+  - [L2StandardBridge Deployment](#l2standardbridge-deployment)
+  - [L1Block Proxy Update](#l1block-proxy-update)
+  - [L2CrossDomainMessenger Proxy Update](#l2crossdomainmessenger-proxy-update)
+  - [L2StandardBridge Proxy Update](#l2standardbridge-proxy-update)
+- [Selection of `ETHER_TOKEN_ADDRESS`](#selection-of-ether_token_address)
+- [Security Considerations](#security-considerations)
+  - [OptimismPortal Token Allowance](#optimismportal-token-allowance)
+  - [Decimal Scaling](#decimal-scaling)
+  - [Interoperability Support](#interoperability-support)
+  - [Wrapped Ether](#wrapped-ether)
+- [Future Directions](#future-directions)
+  - [Enabling a L2 Native ERC20 Token to Pay for Gas](#enabling-a-l2-native-erc20-token-to-pay-for-gas)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Custom gas token is also known as "custom L2 native asset". It allows for an ERC20 token to be collateralized on L1
+to back the native asset on L2. The native asset is used to buy gas which is used to pay for resources on the network
+and is represented by EVM opcodes such as `CALLVALUE` (`msg.value`).
+
+Both the L1 and L2 smart contract systems MUST be able to introspect on the address of the gas paying token to
+ensure the security of the system. The source of truth for the configuration of the address of the token is
+in the L1 `SystemConfig` smart contract.
+
+## Constants
+
+| Constant          | Value                           | Description |
+| ----------------- | ------------------------------- | ----------- |
+| `ETHER_TOKEN_ADDRESS` |  `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` | Represents ether for gas paying asset |
+| `DEPOSITOR_ACCOUNT` | `address(0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001)` | Account with auth permissions on `L1Block` contract |
+| `GAS_PAYING_TOKEN_SLOT` | `bytes32(uint256(keccak256("opstack.gaspayingtoken")) - 1)` | Storage slot that contains the address and decimals of the gas paying token |
+
+## Properties of a Gas Paying Token
```

Comment:
These properties could be validated automatically in the `superchain-registry` CI checks.

Reply:
This is a great callout, will add a section noting this

Reply:
Added a section on automated verification